src/assignments/yunfeng/lesson_4_answers.clj

```diff
@@ -0,0 +1,138 @@
+;;recogize a card problem
+(def suit-map {:S :spade :H :heart :D :diamond :C :club})
+(def rank-map {:2 0 :3 1 :4 2 :5 3 :6 4 :7 5 :8 6 :9 7 :T 8 :J 9 :Q 10 :K 11 :A 12})
+
+(defn parse-suit
+  "parse the string to get the suit and return it as a keyword"
+  [card-string]
+  (keyword (subs card-string 0 1)))
+
+;;test parse-suit function
+(parse-suit "ST")
+
+(defn parse-rank
+  "parse the string to get the rank and return it as a keyword"
+  [card-string]
+  (keyword (subs card-string 1)))
+
+;;test parse-rank function
+(parse-rank "ST")
+
+
+(defn recogize-a-card
+  "use keywords to find its correspond values and construct the result"
+  [card-string]
+  {:suit ((parse-suit card-string) suit-map) :rank ((parse-rank card-string) rank-map)})
+
+;;test recongize-a-card function
+(recogize-a-card "ST")
+(recogize-a-card "SA")
+(recogize-a-card "H4")
+
+
+
+;;best hand problem
+(defn remove-duplicate-ranks
+  "It is trying to remove duplicate ranks in a hand regardless of what the suit is and return the counted result"
+  [recongized-hand]
+  (->> (map :rank recongized-hand)
+       (set)
+       (count)))
+
+;;test remove-duplicate-ranks function
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 12} {:suit :heart :rank 12} {:suit :diamond :rank 12} {:suit :spade :rank 3}])
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 12} {:suit :heart :rank 12} {:suit :diamond :rank 11} {:suit :spade :rank 3}])
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 12} {:suit :heart :rank 10} {:suit :diamond :rank 11} {:suit :spade :rank 3}])
+(remove-duplicate-ranks [{:suit :spade :rank 12} {:suit :club :rank 9} {:suit :heart :rank 10} {:suit :diamond :rank 11} {:suit :spade :rank 3}])
+
+
+(defn find-duplicate-ranks
+  "It is trying to find duplicate ranks regardless of what the suit is"
+  [recongized-hand]
+  (->> (map :rank recongized-hand)
+       (frequencies)
+       (filter #(-> % val (> 1)))
+       (keys)))
+
+;;test find-duplicate-ranks function
+(find-duplicate-ranks [{:suit :spade :rank 12} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+(find-duplicate-ranks [{:suit :diamond :rank 0} {:suit :spade :rank 0} {:suit :spade :rank 1} {:suit :spade :rank 2} {:suit :spade :rank 3}])
+
+
+(defn flush?
+  "Check if a hand is a flush"
+  [recongized-hand]
```

Comment:
You have this interesting typo throughout: recongized for recognized :-)

Reply:
haha. I forgot to check the typo